aten/src/ATen/native/cpu/SumKernel.cpp

```diff
@@ -0,0 +1,304 @@
+#include <ATen/Dispatch.h>
+#include <ATen/native/TensorIterator.h>
+#include <ATen/native/ReduceOps.h>
+#include <ATen/native/cpu/Reduce.h>
+#include <c10/util/llvmMathExtras.h>
+
+#include <algorithm>
+
+
+namespace at {
+namespace native {
+namespace {
+
+template <typename scalar_t>
+struct LoadImpl {
+  static scalar_t load(const char * C10_RESTRICT data, int64_t stride, int64_t index) {
+    auto *ptr = reinterpret_cast<const scalar_t*>(data + index * stride);
+    return *ptr;
+  }
+};
+
+template <typename scalar_t>
+struct LoadImpl<Vec256<scalar_t>> {
+  static Vec256<scalar_t> load(const char * C10_RESTRICT data, int64_t stride, int64_t index) {
+    auto *ptr = data + index * stride;
+    return Vec256<scalar_t>::loadu(ptr);
+  }
+};
+
+template <typename T>
+T load(const char * C10_RESTRICT data, int64_t stride, int64_t index) {
+  return LoadImpl<T>::load(data, stride, index);
+}
+
+template <typename scalar_t>
+void store_result(char * C10_RESTRICT data, int64_t stride, int64_t index, scalar_t value) {
+  auto * ptr = reinterpret_cast<scalar_t*>(data + index * stride);
+  *ptr += value;
+}
+
+template <typename scalar_t, size_t numel>
+void store_result(char * C10_RESTRICT data, int64_t stride, int64_t index,
+    const std::array<scalar_t, numel> &values) {
+  auto *base_ptr = data + stride * index;
+  for (int64_t k = 0; k < numel; ++k) {
+    store_result(base_ptr, stride, k, values[k]);
+  }
+}
+
+int64_t ceil_log2(int64_t x) {
+  if (x <= 2) {
+    return 1;
+  }
+
+  auto ux = static_cast<uint64_t>(x);
+  // Last set bit is floor(log2(x)), floor + 1 is ceil
+  // except when x is an exact powers of 2, so subtract 1 first
+  return static_cast<int64_t>(llvm::findLastSet(ux - 1)) + 1;
+}
+
+// Simultaneously sum over n rows at once
+template <typename scalar_t, int64_t nrows>
+std::array<scalar_t, nrows> multi_row_sum(
+    const char * C10_RESTRICT in_data,
+    const int64_t row_stride,
+    const int64_t col_stride,
+    const int64_t size) {
+  constexpr int64_t num_levels = 4;
```

Comment:
This is a tunable parameter, right?

Reply:
Yes, adding more accumulator levels improves accuracy for larger arrays but may increase register pressure.